app/gateway/api/handler/delivery_handler.go

```diff
@@ -63,3 +69,100 @@ func (h *Handler) ListDeliveries() http.HandlerFunc {
 		rest.SendJSON(rw, resp.Status, resp.Payload, resp.Headers) //nolint:errcheck
 	}
 }
+
+func (h *Handler) CreateDelivery() http.HandlerFunc {
+	return func(rw http.ResponseWriter, req *http.Request) {
+		deliveryBody := &schema.CreateDeliveryRequest{}
+		err := json.NewDecoder(req.Body).Decode(deliveryBody)
+
+		var resp *response.Response
+
+		if err != nil {
+			resp = response.InternalServerError(err)
+			rest.SendJSON(rw, resp.Status, resp.Payload, resp.Headers) //nolint:errcheck
+
+			return
+		}
+
+		err = schema.ValidateCreateDeliveryRequest(deliveryBody)
+		if err != nil {
+			resp := response.BadRequest(err, err.Error())
+			rest.SendJSON(rw, resp.Status, resp.Payload, resp.Headers) //nolint:errcheck
+
+			return
+		}
+
+		useCaseInput := usecase.CreateDeliveryInput{}
+		useCaseInput.Delivery.Qty = deliveryBody.Qty
+		useCaseInput.Delivery.MedicineID = deliveryBody.MedicineID
+		useCaseInput.Delivery.UnitID = deliveryBody.UnitID
+		useCaseInput.Delivery.ClientID = deliveryBody.ClientID
+
+		reference := strconv.Itoa(rand.Intn(schema.MaxReference) + schema.MinReference) //nolint:gosec
+		_, err = h.useCase.GetDeliveryByReference(req.Context(), usecase.GetDeliveryByReferenceInput{
+			Reference: reference,
+		})
+
+		for err == nil {
+			reference = strconv.Itoa(rand.Intn(schema.MaxReference) + schema.MinReference) //nolint:gosec
+
+			_, err = h.useCase.GetDeliveryByReference(req.Context(), usecase.GetDeliveryByReferenceInput{
+				Reference: reference,
+			})
+		}
+
+		if strings.Contains(err.Error(), "no rows in result set") {
+			useCaseInput.Delivery.Reference = reference
+		} else {
+			resp := response.InternalServerError(err)
+			rest.SendJSON(rw, resp.Status, resp.Payload, resp.Headers) //nolint:errcheck
+
+			return
+		}
```

Comment:
```suggestion
		for {
			reference := strconv.Itoa(rand.Intn(schema.MaxReference) + schema.MinReference) //nolint:gosec
			_, err = h.useCase.GetDeliveryByReference(req.Context(), usecase.GetDeliveryByReferenceInput{
				Reference: reference,
			})

			if err != nil {
				if strings.Contains(err.Error(), "no rows in result set") {
					useCaseInput.Delivery.Reference = reference
					break
				}

				resp := response.InternalServerError(err)
				rest.SendJSON(rw, resp.Status, resp.Payload, resp.Headers) //nolint:errcheck

				return
			}
		}
```
O ideal seria fazer um for aqui para ficar retentando gerar até conseguir uma referência válida. Da forma que você fez, ele vai retentar apenas uma vez